pennylane/transforms/qcut.py

```diff
@@ -210,3 +213,235 @@ def tape_to_graph(tape: QuantumTape) -> MultiDiGraph:
             order += 1
 
     return graph
+
+
+@dataclass(frozen=True)
+class CutConfig:
+    """
+    Container wrapping partitioning parameters for passing to the automatic cutter.
+    """
+
+    k: int
+    imbalance: float
+    fragment_wires: List[int] = None
+    fragment_gates: List[int] = None
+
+
+def _infer_cut_configs(
+    num_tape_wires,
+    num_tape_gates,
+    max_device_wires,
+    max_device_gates,
+    min_device_wires=None,
+    min_device_gates=None,
+    k_lower=None,
+    k_upper=None,
+    imbalance_tolerance=None,
+    fragment_wires=None,
+    fragment_gates=None,
+):
+    """
+    Helper function for deriving the best default partitioning parameters for the automatic
+    circuit cutter. Arguments can be either user provided or inferred from devices.
+
+    Args:
+        num_tape_wires (int): Number of wires in the circuit tape to be partitioned.
+        num_tape_gates (int): Number of gates in the circuit tape to be partitioned.
+        max_device_wires (int): Number of wires for the largest available device.
+        max_device_gates (int): Maximum allowed circuit depth for the deepest available device.
+        min_device_wires (int): Number of wires for the smallest available device.
+            Optional, defaults to `max_device_wires`.
+        min_device_gates (int): Maximum allowed circuit depth for the shallowest available device.
+            Optional, defaults to `max_device_gates`.
+        k_lower (int): Lowest number of fragments to attempt for the automatic circuit cutter.
+            Optional, defaults to the scenario where each fragment is executed on the largest
+            available device, satisfying both wire and gate constraints.
```

Comment:
The extreme cases here are solely for the calculation of `k`'s bounds and has nothing to do with how we run them. It doesn't mean we will carry out the execution, nor does it guarantee the petitioner will come up with fragments to realize these extreme scenarios. In other words, it's the best/worst scenarios in terms of number of fragments required for running a tape on a pool of devices.

Take your example of devices with 4, 5, and 6 wires:
- if the circuit has 4 wires, `k_lower` and `k_upper` both defaults to 1 since any device suffices;
- if the circuit has 5 wires, `k_lower` defaults to 1 since it can be run on the 5- or 6-wire devices, but `k_upper` defaults to 2 since, if a user requires, two 4-wire devices will be required. Of course `k` can be higher than `k_upper` (e.g. running a (2,2,1) split on three 4-wire devices), but that's for the user to decide if they want to explore the option.
- if the circuit has 11 wires, you'll need at least two 6- or 5- wire devices, but three 4-wire devices. So `k_lower` defaults to 2 since this is the least number of cuts we'll need in order to fit things onto the available devices. `k_upper` defaults to 3 in the scenario we only use the 4-wire one. But users get to decide if they want to do more exploration by setting it higher than 3. For example, if the circuit is in fact itself already completely disconnected in to 4 parts.